File: src/test/java/testmaintenanceexercises/UnitTests.java
package testmaintenanceexercises;

import static org.junit.Assert.*;

import org.junit.Ignore;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.transaction.annotation.Isolation;


public class UnitTests {
	@Test
	public void CancelTheFirstNumber() {
		Calculator c = new Calculator();
		c.press("1");
		c.press("C");
		String result = c.getDisplay();
		assertEquals("0", result);
	}

	@Test
	public void CancelTheFirstNumber2() {
		Calculator c = new Calculator();
		c.press("1");
		c.press("C");
		c.press("2");
		String result = c.getDisplay();
		assertEquals("2", result);
	}

	@Test
	@Category(Isolation.class)
	public void displayOnlyIfUserIsLoggedIn(){
		Calculator calc = new Calculator();
		if (calc.getCurrentUser() != null)
			assertEquals("0",calc.getDisplay());
	}
	
	@Value("memory")
	private String memory;
	
	@Test
	public void restoreMemoryAndContinue() {
		Calculator calculator = new Calculator();
		String lastStoredValue = calculator.getLastValueFor("Gil");
		if (lastStoredValue == memory) {
			calculator.press("2");
			String result = calculator.getDisplay();
			assertEquals("12", result);
		}
	}

	@Test
	public void CancelAfterOperation() {
		String result; 
		Calculator c = new Calculator();
		c.press("1");
		c.press("+");
		c.press("C");
		result = c.getDisplay();
	}

	@Test
 	public void NotEnoughInfo() {
		Calculator c = new Calculator();
		c.press("1");
		String result = c.getDisplay();
		assertTrue(result.equals("1"));
	}

	@Test
	public void RedundnatTest1() {
		Calculator c = new Calculator();
		c.press("1");
		c.press("2");
		String result = c.getDisplay();
		assertEquals(result, "12");
	}

	@Test
	public void RedundnatTest2() {
		Calculator c = new Calculator();
		c.press("9");
		c.press("5");
		String result = c.getDisplay();
		assertEquals(result, "95");
	}

	@Ignore
	@Test
	public void RedundnatTest4() {
		Calculator c = new Calculator();
		c.press("7");
		c.press("3");
		String result = c.getDisplay();
		assertEquals(result, "73");
	}
	
	@Test
	public void orderOfOperations() {
		/// This is a long test
		Calculator c = new Calculator();
		c.press("1");
		c.press("+");
		c.press("2");
		c.press("*");
		c.press("4");
		c.press("=");
		String result = c.getDisplay();
		assertEquals(result, "9");
	}

}

File: CalculatorServerTests/UnitTests/UnitTests.cs
using CalculatorServer;
using System;
using Xunit;

namespace CalculatorServerTests
{
    public class UnitTests
    {
        public UnitTests()
        {
            Calculator.Reset();
        }
        [Fact]
        public void CancelTheFirstNumber()
        {
            Calculator c = new Calculator(null);
            c.Press("1");
            c.Press("C");
            String result = c.Display;
            Assert.Equal("0", result);
        }

        [Fact]
        public void CancelTheFirstNumber2()
        {
            Calculator c = new Calculator(null);
            c.Press("1");
            c.Press("C");
            c.Press("2");
            String result = c.Display;
            Assert.Equal("2", result);
        }

        [Fact]
        public void DisplayOnlyIfUserIsLoggedIn()
        {
            Calculator calc = new Calculator(null);
            if (String.IsNullOrEmpty(calc.CurrentUser))
                Assert.Equal("0", calc.Display);
        }

        [Fact(Skip ="Need to mock DbContext")]
        public void RestoreMemoryAndContinue()
        {

            MockUserDbContext mockUserRepository = new MockUserDbContext();
            User mockUser = new User();
            mockUser.Name = "Gil";
            mockUser.Memory = 2;
            mockUserRepository.mockUser = mockUser;
            Calculator calculator = new Calculator(mockUserRepository);
            calculator.GetLastValueFor("Gil");
            String lastStoredValue = calculator.Display;
            if (lastStoredValue.Equals("2"))
            {
                calculator.Press("3");
                String result = calculator.Display;
                Assert.Equal("23", result);
            }
        }

        [Fact]
        public void CancelAfterOperation()
        {
            String result;
            Calculator c = new Calculator(null);
            c.Press("1");
            c.Press("+");
            c.Press("C");
            result = c.Display;
        }

        [Fact]
        public void OnePress()
        {
            Calculator c = new Calculator(null);
            c.Press("1");
            String result = c.Display;
            Assert.True(result == "1");
        }

        [Fact]
        public void TwoNumbers()
        {
            Calculator c = new Calculator(null);
            c.Press("1");
            c.Press("2");
            String result = c.Display;
            Assert.Equal(result, "12");
        }

        [Fact]
        public void AnotherTwoNumbers()
        {
            Calculator c = new Calculator(null);
            c.Press("9");
            c.Press("5");
            String result = c.Display;
            Assert.Equal(result, "95");
        }

        [Fact(Skip = "Whatever")]
        public void OtherTwoNumbers()
        {
            Calculator c = new Calculator(null);
            c.Press("7");
            c.Press("3");
            String result = c.Display;
            Assert.Equal(result, "73");
        }

        [Fact(Skip = "I'll get there")]
        public void orderOfOperations()
        {
            Calculator c = new Calculator(null);
            c.Press("1");
            c.Press("+");
            c.Press("8");
            c.Press("/");
            c.Press("4");
            c.Press("=");
            String result = c.Display;
            Assert.Equal(result, "3");
        }

    }
}

